include/ametsuchi/buffer.h

```diff
@@ -0,0 +1,135 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2017 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <ametsuchi/exception.h>
+#include <ametsuchi/globals.h>
+#include <cmath>
+
+namespace ametsuchi {
+namespace buffer {
+
+/**
+ * Basic circular buffer
+ */
+template <typename T>
+class CircularStack {
+ public:
+  CircularStack(size_t);
+  CircularStack(CircularStack &&);
+  CircularStack(const CircularStack &) = delete;
+  ~CircularStack();
+
+  /**
+   * Performs reallocation of the data
+   * so the buffer able to store provided
+   * number of elems
+   * O(n)
+   */
+  void resize(size_t);
+
+  /**
+   * O(1)
+   * Add element to the buffer end pointer
+   */
+  void push(T &&);
+
+  /**
+   * O(1)
+   * Add element to the buffer end pointer
+   */
+  void push(const T &);
+  /**
+   * O(1)
+   * @param n elemens for removing
+   */
+  void pop(size_t n = 1);
+
+  /**
+   * O(1)
+   * Accessing element by index
+   */
+  T &operator[](size_t);
+  T operator[](size_t) const;
+
+  /**
+   * Number of elements
+   */
+  inline size_t size() const { return sz; }
```

Comment:
Does is make sense to use `inline`? Compilers generally ignore it and inline where necessary.